shared/src/balancer/logexpmath.rs

```diff
@@ -0,0 +1,522 @@
+// https://github.com/balancer-labs/balancer-v2-monorepo/blob/6c9e24e22d0c46cca6dd15861d3d33da61a60b98/pkg/solidity-utils/contracts/math/LogExpMath.sol
+
+#![allow(dead_code)]
+
+use ethcontract::{I256, U256};
+use lazy_static::lazy_static;
+use std::convert::TryFrom;
+
+lazy_static! {
+    static ref ONE_18: I256 = I256::exp10(18);
+    static ref ONE_20: I256 = I256::exp10(20);
+    static ref ONE_36: I256 = I256::exp10(36);
+    static ref UNSIGNED_ONE_18: U256 = U256::try_from(*ONE_18).unwrap();
+    static ref MAX_NATURAL_EXPONENT: I256 = ONE_18.checked_mul(I256::from(130_i128)).unwrap();
+    static ref MIN_NATURAL_EXPONENT: I256 = ONE_18.checked_mul(I256::from(-41_i128)).unwrap();
+    static ref LN_36_LOWER_BOUND: I256 = ONE_18.checked_sub(I256::exp10(17)).unwrap();
+    static ref LN_36_UPPER_BOUND: I256 = ONE_18.checked_add(I256::exp10(17)).unwrap();
+    static ref MILD_EXPONENT_BOUND: U256 = (U256::one() << 254_u32)
+        .checked_div(U256::try_from(*ONE_20).unwrap())
+        .unwrap();
+}
+fn constant_x_20(i: u32) -> I256 {
+    match i {
+        2 => 3_200_000_000_000_000_000_000_i128,
+        3 => 1_600_000_000_000_000_000_000_i128,
+        4 => 800_000_000_000_000_000_000_i128,
+        5 => 400_000_000_000_000_000_000_i128,
+        6 => 200_000_000_000_000_000_000_i128,
+        7 => 100_000_000_000_000_000_000_i128,
+        8 => 50_000_000_000_000_000_000_i128,
+        9 => 25_000_000_000_000_000_000_i128,
+        10 => 12_500_000_000_000_000_000_i128,
+        11 => 6_250_000_000_000_000_000_i128,
+        _ => panic!("Constant not provided"),
+    }
+    .into()
+}
+fn constant_x_18(i: u32) -> I256 {
+    match i {
+        0 => 128_000_000_000_000_000_000_i128,
+        1 => 64_000_000_000_000_000_000_i128,
+        _ => panic!("Constant not provided"),
+    }
+    .into()
+}
+fn constant_a_20(i: u32) -> I256 {
+    match i {
+        2 => 7_896_296_018_268_069_516_100_000_000_000_000_i128,
+        3 => 888_611_052_050_787_263_676_000_000_i128,
+        4 => 298_095_798_704_172_827_474_000_i128,
+        5 => 5_459_815_003_314_423_907_810_i128,
+        6 => 738_905_609_893_065_022_723_i128,
+        7 => 271_828_182_845_904_523_536_i128,
+        8 => 164_872_127_070_012_814_685_i128,
+        9 => 128_402_541_668_774_148_407_i128,
+        10 => 113_314_845_306_682_631_683_i128,
+        11 => 106_449_445_891_785_942_956_i128,
+        _ => panic!("Constant not provided"),
+    }
+    .into()
+}
+fn constant_a_18(i: u32) -> I256 {
+    match i {
+        0 => {
+            I256::from_dec_str("38877084059945950922200000000000000000000000000000000000").unwrap()
+        }
+        1 => 6_235_149_080_811_616_882_910_000_000_i128.into(),
+        _ => panic!("Constant not provided"),
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Errors {
+    XOutOfBounds,
+    YOutOfBounds,
+    ProductOutOfBounds,
+    InvalidExponent,
+}
+
+pub fn pow(x: U256, y: U256) -> Result<U256, Errors> {
```

Comment:
These are technically `fixed128x18` right? I know the code doesn't use this, but it might make sense to add a type alias like:
```rust
type Fixed128x18 = U256
```

Just so that its more obvious what the "unit" of the numbers are?

Reply:
Makes sense! Just to be sure since I'm not familiar with the notation, you mean `fixed256x18`, right?

Reply:
I actually meant `Ufixed128x18` which is the [default fixed point size in Solidity](https://docs.soliditylang.org/en/v0.8.5/types.html#fixed-point-numbers).

The notation `ufixedMxN` implies:
- M is the number of bits of the integer part of the number number
- N is the decimals in 10^N

It can't be `fixed256x18` as that would imply that there are 256 bits for the integer which would leave no bits for the fractional part of the number.

Note that this notation is very strange to me and very particular to Solidity. Typically, fixed point numbers are denoted as `NxF` where both N and F are the number of bits of the integer and fractional components respectively. Solidity decided that the fractional "size" for fixed point numbers should be in base 10 instead of 2 - which is a little odd.

Reply:
> M is the number of bits of the integer part of the number number

Then it should be `ufixed197x18`, where `197 = ceil(log2(2^256÷10^18))`, right?
In the link I read:

> M represents the number of bits taken by the type

which is why I'd say 256.

Reply:
> > M represents the number of bits taken by the type
>
> which is why I'd say 256.

Ah right! I think you might be right... I'm not sure because Solidity fixed point numbers confuse me :confounded:.